File: src/proxy/stats.rs
use std::sync::atomic::{AtomicI64, AtomicU64, Ordering};
use std::fmt;

pub struct Stats {
    pub connections_total: AtomicU64,
    pub connections_active: AtomicI64,
    pub connections_ws: AtomicU64,
    pub connections_tcp_fallback: AtomicU64,
    pub connections_cfproxy: AtomicU64,
    pub connections_bad: AtomicU64,
    pub connections_masked: AtomicU64,
    pub ws_errors: AtomicU64,
    pub bytes_up: AtomicU64,
    pub bytes_down: AtomicU64,
    pub pool_hits: AtomicU64,
    pub pool_misses: AtomicU64,
    pub cf_pool_hits: AtomicU64,
    pub cf_pool_misses: AtomicU64,
}

impl Default for Stats {
    fn default() -> Self {
        Self::new()
    }
}

impl Stats {
    pub const fn new() -> Self {
        Self {
            connections_total: AtomicU64::new(0),
            connections_active: AtomicI64::new(0),
            connections_ws: AtomicU64::new(0),
            connections_tcp_fallback: AtomicU64::new(0),
            connections_cfproxy: AtomicU64::new(0),
            connections_bad: AtomicU64::new(0),
            connections_masked: AtomicU64::new(0),
            ws_errors: AtomicU64::new(0),
            bytes_up: AtomicU64::new(0),
            bytes_down: AtomicU64::new(0),
            pool_hits: AtomicU64::new(0),
            pool_misses: AtomicU64::new(0),
            cf_pool_hits: AtomicU64::new(0),
            cf_pool_misses: AtomicU64::new(0),
        }
    }

    pub fn summary(&self) -> String {
        let pool_total = self.pool_hits.load(Ordering::Relaxed)
            + self.pool_misses.load(Ordering::Relaxed);
        let pool_s = if pool_total > 0 {
            format!(
                "{}/{}",
                self.pool_hits.load(Ordering::Relaxed),
                pool_total
            )
        } else {
            "n/a".into()
        };
        let cf_pool_total = self.cf_pool_hits.load(Ordering::Relaxed)
            + self.cf_pool_misses.load(Ordering::Relaxed);
        let cf_pool_s = if cf_pool_total > 0 {
            format!(
                "{}/{}",
                self.cf_pool_hits.load(Ordering::Relaxed),
                cf_pool_total
            )
        } else {
            "n/a".into()
        };
        format!(
            "total={} active={} ws={} tcp_fb={} cf={} bad={} masked={} err={} pool={} cf_pool={} up={} down={}",
            self.connections_total.load(Ordering::Relaxed),
            self.connections_active.load(Ordering::Relaxed),
            self.connections_ws.load(Ordering::Relaxed),
            self.connections_tcp_fallback.load(Ordering::Relaxed),
            self.connections_cfproxy.load(Ordering::Relaxed),
            self.connections_bad.load(Ordering::Relaxed),
            self.connections_masked.load(Ordering::Relaxed),
            self.ws_errors.load(Ordering::Relaxed),
            pool_s,
            cf_pool_s,
            human_bytes(self.bytes_up.load(Ordering::Relaxed)),
            human_bytes(self.bytes_down.load(Ordering::Relaxed)),
        )
    }
}

pub fn human_bytes(n: u64) -> String {
    let units = ["B", "KB", "MB", "GB", "TB"];
    let mut size = n as f64;
    for unit in &units {
        if size < 1024.0 {
            return format!("{:.1}{}", size, unit);
        }
        size /= 1024.0;
    }
    format!("{:.1}TB", size)
}

impl fmt::Display for Stats {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.summary())
    }
}

File: proxy/stats.py
import time
from .utils import human_bytes


class _Stats:
    def __init__(self):
        self.connections_total = 0
        self.connections_active = 0
        self.connections_ws = 0
        self.connections_tcp_fallback = 0
        self.connections_cfproxy = 0
        self.connections_bad = 0
        self.connections_masked = 0
        self.ws_errors = 0
        self.bytes_up = 0
        self.bytes_down = 0
        self.pool_hits = 0
        self.pool_misses = 0
        self.cf_pool_hits = 0
        self.cf_pool_misses = 0
        self._start_time: float = time.monotonic()
        # per-DC counters: {dc_key: count} e.g. {'2': 5, '4m': 2}
        self.dc_connections: dict = {}

    def reset_counters(self) -> None:
        """Reset all cumulative counters (but keep connections_active)."""
        active = self.connections_active
        self.__init__()
        self.connections_active = active

    def record_dc(self, dc: int, is_media: bool) -> None:
        key = f"{dc}{'m' if is_media else ''}"
        self.dc_connections[key] = self.dc_connections.get(key, 0) + 1

    def uptime_seconds(self) -> float:
        return time.monotonic() - self._start_time

    def uptime_str(self) -> str:
        total = int(self.uptime_seconds())
        h, rem = divmod(total, 3600)
        m, s = divmod(rem, 60)
        if h:
            return f"{h}ч {m:02d}м {s:02d}с"
        if m:
            return f"{m}м {s:02d}с"
        return f"{s}с"

    def tray_tooltip(self) -> str:
        """Short one-line summary for the tray icon tooltip."""
        return (
            f"TG WS Proxy  |  "
            f"активных: {self.connections_active}  "
            f"всего: {self.connections_total}  "
            f"↑{human_bytes(self.bytes_up)} ↓{human_bytes(self.bytes_down)}"
        )

    def summary(self) -> str:
        pool_total = self.pool_hits + self.pool_misses
        pool_s = (f"{self.pool_hits}/{pool_total}"
                  if pool_total else "n/a")
        cf_pool_total = self.cf_pool_hits + self.cf_pool_misses
        cf_pool_s = (f"{self.cf_pool_hits}/{cf_pool_total}"
                     if cf_pool_total else "n/a")
        return (f"total={self.connections_total} "
                f"active={self.connections_active} "
                f"ws={self.connections_ws} "
                f"tcp_fb={self.connections_tcp_fallback} "
                f"cf={self.connections_cfproxy} "
                f"bad={self.connections_bad} "
                f"masked={self.connections_masked} "
                f"err={self.ws_errors} "
                f"pool={pool_s} "
                f"cf_pool={cf_pool_s} "
                f"up={human_bytes(self.bytes_up)} "
                f"down={human_bytes(self.bytes_down)}")


stats = _Stats()